File: Data-Structures/Stack/StackUsingLL.java
import java.util.Scanner;

public class StackUsingLL {
    Node head;
    class Node {
        int val;
        Node next;
        int min;

    public Node(int val, int min) {
        this.val = val;
        this.min = min;
        next = null;
        }
    }

    public void push(int x) { // O(1)
        if(head == null){
            head = new Node(x, x);
        }
        else {
            Node node = new Node(x, Math.min(head.min,x));
            node.next = head;
            head = node;
            }
        }

        public void pop() {
        if(head!=null){
            head = head.next;
            }
        }

        public int peek() {
        if(head!=null) {
            return head.val;
            }
        return -1;
        }

        public int getMin() {
        if(head!=null){
            return head.min;
        }
        return -1;
        }

        public void display() {

            // check for stack underflow
            if (head == null) {
                System.out.printf("\nStack Underflow");
            }
            else {
                Node temp = head;
                while (temp != null) {
                    System.out.print(temp.val+" ");
                    temp = temp.next;
                }
                System.out.println("");
            }
        }
        
    public static void operation(StackUsingLL stack) {
    	 System.out.println("Enter the operation");
         System.out.println("1. "+"push");
         System.out.println("2. pop");
         System.out.println("3. peek");
         System.out.println("4. display");
         
         Scanner sc = new Scanner(System.in);
         int input = sc.nextInt();
         
         switch(input) {
         // push into stack
         case 1:
             System.out.println("Enter the number of elements to be pushed");
             int number = sc.nextInt();
             
             System.out.println("Enter the elements");
             while(number-->0) {
             	int element = sc.nextInt();
             	stack.push(element);
             }
             stack.operation(stack);
          // Delete top element of Stack
           case 2:
         	stack.pop();
         	stack.operation(stack);
             
          // print Top element of Stack  
           case 3:
         	  System.out.printf("\nTop element is %d\n", stack.peek());
         	 stack.operation(stack);
             
           //print Stack elements  
           case 4:
         	stack.display();
         	stack.operation(stack);
         	
           case 5:
        	System.out.println("EXIT");
        	System.exit(0);
         	
           default: System.exit(0);
         }
    }

    public static void main(String[] args) {
        //object of class
    	StackUsingLL stack = new StackUsingLL();
        stack.operation(stack);
    }
}

File: Data-Structures/Stack/stackusingLL.c
#include <stdio.h>
#include <stdlib.h>
struct node{
  int data;
  struct node *next;
};
struct stack{
  int top;
  struct node *list;//head of list
};
struct node *newNode(){
  struct node *newnode=(struct node*)malloc(sizeof(struct node));
  printf("\nEnter the number to be inserted\n");
  scanf("%d",&newnode->data);
  newnode->next=NULL;
  return newnode;
}
struct stack *create_stack(){
  struct stack *st=(struct stack*)malloc(sizeof(struct stack));
  st->list=NULL;
  st->top=-1;
  return st;
}
void push(struct node **head){
  struct node *n=newNode();
  if(*head==NULL){
    *head=n;
  }
  else{
    struct node *p=*head;
    n->next=p;
    *head=n;
  }
}
void pop(struct node **head){
  if(*head==NULL){
    printf("\nUNDERFLOW\n");
  }
  else{
    struct node *temp=*head;
    *head=temp->next;
    printf("\nDeleted %d\n",temp->data);
    free(temp);
  }
}
void top_element(struct node *head){
  printf("\nTop element is %d\n",head->data);
}
void print_stack(struct node *head){
  printf("\n");
  struct node *p=head;
  while(p!=NULL){
    printf("%d ",p->data);
    p=p->next;
  }
  printf("\n");
}
int main(){
  struct stack *st=create_stack();
  char ch;
  printf("a->push\nd->pop\np->find top element\nk->print stack\nq->quit\n");
  do{
    printf("\nEnter a character from the given menu to perform the operations on stack\n");
    ch=getche();
    switch(ch){
      case 'a': ;
      push(&st->list);
      break;
      case 'd': ;
      pop(&st->list);
      break;
      case 'p': ;
      top_element(st->list);
      break;
      case 'q': ;
      break;
      case 'k': ;
      print_stack(st->list);
      break;
      default:
      printf("\nIllegal Character entered\n");
      break;
    }
  }while(ch!='q');
}
